### PR TITLE
[RAC-5256]Deploy vNode failure should be captured in test scripts 

### PR DESCRIPTION
**Pull request by HQebupt**
## 1. Backgroud
There should be 3 vNodes deployed.
But there are only 2 vNodes deployed and the shell step didn't exit with error code.

`Error: Duplicate name 'vmslave16-Rinjin1' on target. Use --overwrite option to delete it.
Completed with errors`

Why do the jenkins can not capture the ERROR ? 

## 2.  Analysis
The key cause is the command: "expect" in the method `execWithTimeout() ` of `shareMethod.sh`.
`expect -c 'set timeout 90; spawn -noecho /bin/sh -c "ovftool --noSSLVerify --diskMode= --datastore=Isilon-Cluster-NFS  --name='\''vmslave16-Rinjin1'\'' --net:'\''Mgmt - longej=vmslave16-switch'\'' '\''/home/****/isofarm/OVA/vRinjin-Haswell.ova'\''   vi://****:****@10.240.19.230"; expect timeout { exit 1 } eof { exit 0 }'`

- if `ovftool` executes timeout, return 1.
- if `ovftool` occurrs other error, such as "Duplicate name 'vmslave16-Rinjin1' on target.", the command expect returns 0. Therefore the Jenkins can not capture this error.(That is unexpected.)

Generally, the normal behavior of `execWithTimeout()` is
- if `ovftool` executes timeout, return 1.
- else return  what the `ovftool` returns.

@PengTian0 @PengTian0 @anhou @iceiilin @changev

### CODE REVIEW

File: shareMethod.sh [hunk 1]
```diff
@@ -9,9 +9,9 @@ execWithTimeout() {
         echo "execWithTimeout() Command not specified"
         exit 2
     fi
-    local cmd="/bin/sh -c \"$1\""
+    export cmd="/bin/sh -c \"$1\""
```
Comment (PengTian0) [inline shareMethod.sh]: Why change to ```export```?

Comment (HQebupt) [reply shareMethod.sh]: The `expect << EOF` can not recognize the local variable.  I test several ways, and it is a unique way to do so.

Comment (panpan0000) [reply shareMethod.sh]: I'm afraid ```export``` will bring variable naming pollution ..
and I didn't understand what do you mean. if EOF not working ,why not expect -c?

File: shareMethod.sh [hunk 2]
```diff
@@ -25,7 +25,16 @@ execWithTimeout() {
...
     do
-        expect -c "set timeout $timeout; spawn -noecho $cmd; expect timeout { exit 1 } eof { exit 0 }"
+        expect << 'EOF'
+            set timeout $env(timeout)
+            eval spawn -noecho $env(cmd)
+            expect {
+                timeout { exit 1 }
+                eof
+            }
+            lassign [wait] pid spawnid os_error_flag value
```
Comment (panpan0000) [inline shareMethod.sh]: can you explain why add this line can resolve original issue?

File: shareMethod.sh (hunk 1 above)
Comment (anhou) [reply shareMethod.sh]: @PengTian0  @panpan0000  the same question with https://stackoverflow.com/questions/21928226/how-to-pass-shell-script-variable-into-expect-script  this link could help you understand the background

File: shareMethod.sh (hunk 2 above)
Comment (HQebupt) [reply shareMethod.sh]: `lassign  [wait] pid spawnid os_error_flag value` can get the exit code of spawned process in expect shell script. And the original `expect -c "set timeout $timeout; spawn -noecho $cmd; expect timeout { exit 1 } eof { exit 0 } ` can not, which always returns 0 when `spawn -noecho $cmd` failed.

**Review by panpan0000 — APPROVED** [review]
@PengTian0 to review

**Comment by PengTian0** [comment]
:+1:

**Review by anhou — APPROVED** [review]

**Pull request merged**